Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2021, timmot <tiwwot@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "DownloadWindow.h"
+#include <AK/Debug.h>
+#include <FileSystemAccessServer/ClientConnection.h>
+#include <LibC/fcntl.h>
+#include <LibCore/File.h>
+#include <LibCore/StandardPaths.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/MessageBox.h>
+
+namespace FileSystemAccessServer {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
+    : IPC::ClientConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    GUI::Application::the()->quit();
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileResponse ClientConnection::request_file(String const& sender, String const& path, int requested_access)
+{
+    // TODO: This hangs the calling application
+
+    // Must receive an absolute path
+    VERIFY(path.starts_with("/"sv));
+
+    bool approved = false;
+    if (m_approved_files.get(path).has_value()) {
+        approved = (requested_access & m_approved_files.get(path).value()) == requested_access;
+    }
+
+    if (!approved) {
+        StringBuilder builder;
+        if (requested_access & O_RDONLY)
+            builder.append('r');
+
+        if (requested_access & O_WRONLY)
+            builder.append('w');
+
+        if (requested_access & O_CREAT)
+            builder.append('c');
```

Comment:
Since you only support `O_RDONLY`, `O_WRONLY` and `O_CREAT`, you could maybe `VERIFY` that no other permissions were requested, with something like:
```cpp
VERIFY((requested_access & (O_RDONLY | O_WRONLY | O_CREAT)) == requested_access);

Reply:
I don't know if it is useful for a user to know if `Append`, `Truncate`, `MustBeNew`, or `KeepOnExec` is important for the user to note, but it is important for us to have access to these in the code.

Reply:
You're right, those should also be supported, but not shown to the user. So I guess you can disregard my suggestion. :^)